### ISSUE TITLE
resampling with RMS instead of AVERAGE ?

### ISSUE THREAD

**Issue by oarcher** [OSGeo/gdal#3196]
I use gdal from python, using rasterio.
I work on SAR data with SAFE driver, but I think the problem should be the same with RS2 or ALOS driver.

The information stored in the tiff file is a digital number (DN), and the calibration algorithm use DN**2, so using the AVERAGE resampling method introduce an underestimate of the variable (here sigma0) after applying calibration:
```python
gdal_dn = ds.read(out_shape=out_shape, resampling=Resampling.average)
gdal_sigma0 = apply_calibration(gdal_dn**2)
plt.scatter(gdal_sigma0 , ref_sigma0) # ref_sigma0 was computed from dn at full res, so no errors due to mean(dn)
```
![image](https://user-images.githubusercontent.com/4264660/99604289-66a19280-2a05-11eb-85e4-d5ba501f1d2c.png)

(the spread is due to the fact that the size of the resampling windows are not exactly the same)

The underestimate of sigma0 is because
`mean(dn) <= sqrt(mean(dn**2))`

I've tried to modify [GDALResampleChunk32R_AverageT](https://github.com/OSGeo/gdal/blob/48f4517ffa591808ef8ba76b622eabed2d8ec0d9/gdal/gcore/overview.cpp#L561-618) to compute the square root of the square mean, and the result is much better:

![image](https://user-images.githubusercontent.com/4264660/99604221-47a30080-2a05-11eb-858e-d431f1e71d03.png)

But implementing a full SQUAREAVERAGE resampling method is out of my GDAL code knowledge.

GDAL is very fast for resampling, and many science can be done at lower resolution.  It would be fantastic to have an SQUAREAVERAGE method.

**Comment by piyushrpt** [comment]
It might be possible to implement this using VRTs and pixel functions.

DN -> intensity_VRT using pixel functions -> average_VRT using average resampling -> sqrt_VRT using pixel functions -> translate to raster

**Comment by avalentino** [comment]
@oarcher IMHO this is a very interesting use case for SAR users.
The solution proposed by @piyushrpt is very clear and ready to use.
I would be interested to know of you can observe any time and/or quality performance difference between your "custom" implementation and the solution based on pixel functions.
In the case, it could make sense to submit a PR.

**Comment by piyushrpt** [comment]
This PR might provide hints on pathway taken by the averaging operation - https://github.com/OSGeo/gdal/pull/808. If this operator does get approved for inclusion, I would suggest we use "rootmeansquare" or "rms" for the name. I would be interested in the performance of this approach as well.

**Comment by mgoacolou** [comment]
Hi, I already do the work for overviews computation here https://github.com/OSGeo/gdal/pull/1553 but dir not have time to reply to last questions. 
this could be a good reason to finish it and implement for resampling!

**Comment by oarcher** [comment]
I've tried @piyushrpt proposition using VRT and pixel function, but i was not able to make it works.

The problem is that the resampling is called *before* the pixel function.

```python
gdal.GetDriverByName("VRT").CreateCopy('/tmp/safe.vrt', gdal.Open('myfile.SAFE'))
# i've modified /tmp/safe.vrt to add the pixel function to get /tmp/safe_dn2.vrt
dn2 = rasterio.open('/tmp/safe_dn2.vrt')
dn2_arr = dn2.read(out_shape=out_shape, resampling=Resampling.average))
```

so `dn2_arr` is a `dn**2`, but as the resampling occurred before the power function, it's not an rms. 

Is there any way to use VRT and to call resampling *after* the pixel function ?

or perhaps @mgoacolou #1553 (hi neighbor !) is the best way to do it ?

**Comment by piyushrpt** [comment]
Oh .. each step that I mentioned would be its own VRT pointing to the previous one. You wouldnt have to write these to disk if you dont want to and use /vsimem. Here is an old example from isce2 and earlier versions of gdal (this is no longer needed since nodata is now supported for complex data) - but should give you an idea on how to string these virtual functions together.

```python
def multiLookCpx(infile, outfile, fmt='GTiff', xlooks=None, ylooks=None, noData=None, method='average'):
    '''
    infile - Input file to multilook
    outfile - Output file to multilook
    fmt - Output format
    xlooks - Number of looks in x/range direction
    ylooks - Number of looks in y/azimuth direction
    
    
    input cpx file
        |
    2 band real virtual
        |
    2 band real multilooked virtual
        |
    1 band complex virtual
        |
    output cpx file
        
    '''
    sourcexml = '''    <SimpleSource>
      <SourceFilename>{0}</SourceFilename>
      <SourceBand>1</SourceBand>
    </SimpleSource>'''.format(infile)
    
    ds = gdal.Open(infile, gdal.GA_ReadOnly)

    #Input file dimensions
    xSize = ds.RasterXSize
    ySize = ds.RasterYSize

    #Output file dimensions
    outXSize = xSize//xlooks
    outYSize = ySize//ylooks

    #Temporary filenames
    inmemfile = '/vsimem/cpxlooks.2band.vrt'
    inmemfile2 = '/vsimem/cpxlooks.multilooks.2band.vrt'
    inmemfile3 = '/vsimem/cpxlooks.combine.vrt'
    
    ##This is where we convert it to real bands and multilook
    #Create driver
    driver = gdal.GetDriverByName('VRT')
    rivrtds = driver.Create(inmemfile,xSize, ySize, 0)
    
    #Create realband
    options = ['subClass=VRTDerivedRasterBand',
               'PixelFunctionType=real',
               'SourceTransferType=CFloat32']
    rivrtds.AddBand(gdal.GDT_Float32, options)
    rivrtds.GetRasterBand(1).SetMetadata({'source_0' : sourcexml}, 'vrt_sources')
    
    #Create imagband
    options = ['subClass=VRTDerivedRasterBand',
               'PixelFunctionType=imag',
               'SourceTransferType=CFloat32']
    rivrtds.AddBand(gdal.GDT_Float32, options)
    rivrtds.GetRasterBand(2).SetMetadata({'source_0' : sourcexml}, 'vrt_sources')
    
    ##Add projection information
    rivrtds.SetProjection(ds.GetProjection())
    ds = None
    

    ##Set up options for translation
    gdalTranslateOpts = gdal.TranslateOptions(format='VRT', 
                                              width=outXSize, height=outYSize,
                                             srcWin=[0,0,outXSize*xlooks, outYSize*ylooks],
                                             noData=noData, resampleAlg=method)

    #Apply multilooking on real and imag channels
    mlvrtds = gdal.Translate(inmemfile2, rivrtds, options=gdalTranslateOpts)
    rivrtds = None
    mlvrtds = None
        
    #Write from memory to VRT using pixel functions
    mlvrtds = gdal.OpenShared(inmemfile2)
    cpxvrtds = driver.Create(inmemfile3, outXSize, outYSize, 0)
    cpxvrtds.SetProjection(mlvrtds.GetProjection())
    cpxvrtds.SetGeoTransform(mlvrtds.GetGeoTransform())


    options = ['subClass=VRTDerivedRasterBand',
               'pixelFunctionType=complex',
               'sourceTransferType=CFloat32']
    xmltmpl = '''    <SimpleSource>
      <SourceFilename>{0}</SourceFilename>
      <SourceBand>{1}</SourceBand>
    </SimpleSource>'''
    
    md = {'source_0': xmltmpl.format(inmemfile2, 1),
          'source_1': xmltmpl.format(inmemfile2, 2)}

    cpxvrtds.AddBand(gdal.GDT_CFloat32, options)
    cpxvrtds.GetRasterBand(1).SetMetadata(md, 'vrt_sources')
    mlvrtds = None
        
        
    ###Now create copy to format needed
    driver = gdal.GetDriverByName(fmt)
    outds = driver.CreateCopy(outfile, cpxvrtds)
    cpxvrtds = None
    
    outds = None
    gdal.Unlink(inmemfile)
    gdal.Unlink(inmemfile2)
    gdal.Unlink(inmemfile3)```

**Comment by oarcher** [comment]
Thanks @piyushrpt , I've was able to link vrts.

But this solution is slow, it take ~23s, and consume more memory than simple average.

My first approach modifying [GDALResampleChunk32R_AverageT](https://github.com/OSGeo/gdal/blob/48f4517ffa591808ef8ba76b622eabed2d8ec0d9/gdal/gcore/overview.cpp#L561-618)
 replacing `val` by `val * val` was as fast as as the average method (~1.5s), with little memory footprint.

My reference solution using dask took ~10s, and consume mostly all memory and cpus available.

Here is the code:
```python
def gdal_rms(filename, out_shape):

    sourcexml = '''
        <ComplexSource>
            <SourceFilename>{fname}</SourceFilename>
            <SourceBand>{band}</SourceBand>
        </ComplexSource>
        '''

    ds = gdal.Open(filename, gdal.GA_ReadOnly)

    xsize = ds.RasterXSize
    ysize = ds.RasterYSize
    count = ds.RasterCount

    prefix = '/vsmem'
    dn2file = '%s/dn2.vrt' % prefix
    avgfile = '%s/dn2_avg.vrt' % prefix
    datfile = '%s/dn2_avd_dat.vrt' % prefix

    vrt_driver = gdal.GetDriverByName("VRT")
    ds_dn2 = vrt_driver.Create(dn2file, xsize=xsize, ysize=ysize,
                    bands=0)

    #Create square band
    options = [
        'subClass=VRTDerivedRasterBand',
        'PixelFunctionType=intensity',
        'SourceTransferType=UInt32']
    for iband in range(1, ds.RasterCount + 1) :
        ds_dn2.AddBand(gdal.GDT_UInt32, options)
        ds_dn2.GetRasterBand(iband).SetMetadata(
            {'source_0' : sourcexml.format(fname=filename, band=iband)},
            'vrt_sources')

    ds_dn2.SetProjection(ds.GetProjection())

    ds = None

    ##translate ds_dn2, with GRIORA_Average
    gdalTranslateOpts = gdal.TranslateOptions(
        format='VRT',
        width=out_shape[0], height=out_shape[1],
        srcWin=[0,0,xsize, ysize],
        resampleAlg=gdal.gdalconst.GRIORA_Average)

    ds_dn2_average = gdal.Translate(avgfile, ds_dn2, options=gdalTranslateOpts)
    ds_dn2 = None
    ds_dn2_average = None

    #Write from memory to VRT using pixel functions
    ds_dn2_average = gdal.OpenShared(avgfile)
    ds_dn2_average_data = vrt_driver.Create(datfile, out_shape[0], out_shape[1], 0)
ds_dn2_average_data.SetProjection(ds_dn2_average.GetProjection())
ds_dn2_average_data.SetGeoTransform(ds_dn2_average.GetGeoTransform())

    options = ['subClass=VRTDerivedRasterBand',
               'sourceTransferType=UInt32']
    xmltmpl = '''
        <SimpleSource>
            <SourceFilename>{fname}</SourceFilename>
            <SourceBand>{band}</SourceBand>
        </SimpleSource>
        '''

    options = ['subClass=VRTDerivedRasterBand',
               'SourceTransferType=UInt32']
    for iband in range(1, count + 1) :
        ds_dn2_average_data.AddBand(gdal.GDT_UInt32, options)
ds_dn2_average_data.GetRasterBand(iband).SetMetadata({'source_0' : sourcexml.format(fname=filename, band=iband)}, 'vrt_sources')


    dn2_arr = ds_dn2_average.ReadAsArray()
    dn_arr = dn2_arr ** 0.5
    ds_dn2_average_data = None

    gdal.Unlink(dn2file)
    gdal.Unlink(avgfile)
    gdal.Unlink(datfile)

    return dn_arr
```

Perhaps that I have misunderstood something, and may be some optimisations are possible ?

 anyway, #1553 seems to be a fastest and easiest approach for the end user.

**Comment by piyushrpt** [comment]
There is an sqrt pixel function for the last step as well. Is the slowest part, loading the whole image and taking the sqrt? Using the pixel function will reduce memory footprint for sure.

**Comment by oarcher** [comment]
What's take time is `dn2_arr = ds_dn2_average.ReadAsArray()`.
 the final sqrt on `dn2_arr` is done from python side, and it's very fast, because, the array is *already* resampled.

I've also edited my [previous comment](https://github.com/OSGeo/gdal/issues/3196#issuecomment-732031780), because I was using `ComplexSource` (because I've tried [`exponent`](https://gdal.org/drivers/raster/vrt.html#complexsource)) instead of `SimpleSource`, and wrong file name in:
```python
ds_dn2_average_data.GetRasterBand(iband).SetMetadata({'source_0' : sourcexml.format(fname=filename, band=iband)}, 'vrt_sources')
```
it should be:
```python
ds_dn2_average_data.GetRasterBand(iband).SetMetadata({'source_0' : sourcexml.format(fname=avgfile, band=iband)}, 'vrt_sources')
```

This problem doesn't affect the result, but it's little faster (14s vs 23s). not yet the same performance as the builtin average (1.5s)
